### PR TITLE
Restructuring and testing of FsGrid

### PR DESCRIPTION
**Pull request by cscjlan**
This PR contains many internal changes to FsGrid. The main functionality of the structure has not been changed, mainly things have been moved around and made smaller and more self contained.

Main changes include:
- Code was moved from `fsgrid.hpp` at root to `src/grid.hpp`, while `fsgrid.hpp` still exists for single file include purpose
- FsGridTools was changed from a struct with only static functions to a namespace with static functions
- FsGrid doesn't inherit from FsGridTools anymore
- Many tests have been added
- All member variables of FsGrid are private
- Most member variables are const
- FsGrid constructor does nothing, variables are initialized by separate (static) functions
- Unused and unnecessary variables have been removed
- Code has been simplified significantly
- Coordinate related functionality has been moved to a separate structure

These changes will help in the next steps, which will change the functionality of FsGrid towards a `parallel_for` based approach.

### CODE REVIEW

**Comment by lkotipal** [comment]
Splitting the code into multiple headers is probably a good idea, but it also makes the diff completely useless.
> FsGrid constructor does nothing, variables are initialized by separate (static) functions

This almost gave me a heart attack since I thought you implemented multi-stage initialization. Instead you moved the code from the constructor body to static functions called in the initializer list so that's good.

> Code has been simplified significantly

I'm not _entirely_ certain I agree with this. In particular:
- Is there a reason `FSGRID_MPI_CHECK` and such are macros?
- Do variadic templates used here have any benefits over e.g. collections as parameters?

**Comment by cscjlan** [comment]
> Splitting the code into multiple headers is probably a good idea, but it also makes the diff completely useless.

Yes, that is true. Diffs are overall pretty useless for large refactoring.

> > FsGrid constructor does nothing, variables are initialized by separate (static) functions

> This almost gave me a heart attack since I thought you implemented multi-stage initialization. Instead you moved the code from the constructor body to static functions called in the initializer list so that's good.

That was badly worded, but yeah, this PR actually removes the multi-stage initialization that `FsGrid` was (is) using in `dev` (first calling the constructor, then setting `DX`, `DY`, `DZ` and then `physicalGlobalStart`).


>  Is there a reason FSGRID_MPI_CHECK and such are macros?

Less wordy than writing the entire function: `FSGRID_MPI_CHECK(MPI_Call(...), "Error message")` vs `FsGridTools::writeToCerrAndThrowIfFailed(MPI_Call(...) != MPI_SUCCESS, "Error message")`. Since the macro is just a redirection to a function call and does nothing else, I consider it a benign use case of macros. But if you dislike macros, those can be easily replaced with the more wordy function call.

> Do variadic templates used here have any benefits over e.g. collections as parameters?

With a parameter pack, it's trivial to call the assertion function with arbitrary parameters. I also find the error message at the calling site clearer to read, when it's written like this:

```cpp
"An error related to rank ", rank, " with values foo ", foo, " and bar ", bar
```

I suppose it's possible to construct a tuple from the arguments, but this implementation was trivial to write so I went with it.

**Comment by lkotipal** [comment]
> Yes, that is true. Diffs are overall pretty useless for large refactoring.

Fair enough

> Less wordy than writing the entire function: FSGRID_MPI_CHECK(MPI_Call(...), "Error message") vs FsGridTools::writeToCerrAndThrowIfFailed(MPI_Call(...) != MPI_SUCCESS, "Error message"). Since the macro is just a redirection to a function call and does nothing else, I consider it a benign use case of macros. But if you dislike macros, those can be easily replaced with the more wordy function call.

Just rename the function if it's too wordy? You can also refactor it to take `MPI_Status` and check that against `MPI_SUCCESS` like you've done in the macro, since the function isn't used anywhere else.

> With a parameter pack, it's trivial to call the assertion function with arbitrary parameters

Ah, I'm fine with them in the error checks I suppose, I'm thinking more about the coordinate functions where `args` is always going to be some N indices/real numbers and the return value is the same size so those could just be templates with arguments like `std::array<type, N>`. Maybe even just regular old functions with `std::vector<type>`, though I'm not sure if the loss of specialization would affect performance somehow?

**Comment by cscjlan** [comment]
> > Less wordy than writing the entire function: FSGRID_MPI_CHECK(MPI_Call(...), "Error message") vs FsGridTools::writeToCerrAndThrowIfFailed(MPI_Call(...) != MPI_SUCCESS, "Error message"). Since the macro is just a redirection to a function call and does nothing else, I consider it a benign use case of macros. But if you dislike macros, those can be easily replaced with the more wordy function call.
> 
> Just rename the function if it's too wordy? You can also refactor it to take `MPI_Status` and check that against `MPI_SUCCESS` like you've done in the macro, since the function isn't used anywhere else.

It's also used in the `FSGRID_DEBUG_ASSERT` macro, but yeah, I'm not opposed to removing those macros.

> > With a parameter pack, it's trivial to call the assertion function with arbitrary parameters
> 
> Ah, I'm fine with them in the error checks I suppose, I'm thinking more about the coordinate functions where `args` is always going to be some N indices/real numbers and the return value is the same size so those could just be templates with arguments like `std::array<type, N>`. Maybe even just regular old functions with `std::vector<type>`, though I'm not sure if the loss of specialization would affect performance somehow?

Oh, right, you meant those. The reasoning is that those functions redirect to `Coordinates`'s functions, so this way there's no need to change the function in `FsGrid` if it changes in `Coordinates`. But since we plan in the next phase to remove those redirection functions anyway and make `Coordinates` a public variable of `FsGrid`, it doesn't really matter that much.
